### PR TITLE
Allow interface types for injected views

### PR DESCRIPTION
**Pull request by emilsjolander**
This pull request implements what was talked about in issue #212. In summary it allows any injected field to not only be a view subclass but can also be declared with an interface type.

This pull request also adds  a custom error message for when casting injected views fails. I'm currently slightly uncertain if the direction i went with this is good so i'm very open to suggestions. Currently listener parameters are not cast in this way but if this direction looks good i'll update the pull request so that the listener parameters are cast in a similar way.

### CODE REVIEW

File: butterknife/src/main/java/butterknife/internal/ViewInjector.java [hunk 1]
```diff
@@ -129,7 +129,9 @@ private void emitCollectionBinding(StringBuilder builder, CollectionBinding bind
         builder.append(',');
       }
       builder.append("\n        ");
-      emitCastIfNeeded(builder, binding.getType());
+      if (!binding.getType().equals(VIEW_TYPE)) {
```
Comment (JakeWharton) [inline butterknife/src/main/java/butterknife/internal/ViewInjector.java]: Can we extract this onto `binding` as a `requiresCast()` method? I don't like the duplicated logic all over.

**Comment by JakeWharton** [comment]
This looks great. I just have one small suggestion. If you can't get to it in a day or so I'll just make the change myself and merge.

**Comment by emilsjolander** [comment]
@JakeWharton fixed in https://github.com/emilsjolander/butterknife/commit/334000d4df35dec345d94180b66f15364b730a80

**Comment by emilsjolander** [comment]
Added similar casting code for listener parameters. Should be ready to merge now unless you have any comments on the most recent changes.

**Comment by emilsjolander** [comment]
My local setup doesn't seem to test the same tests as travis is set up to do. I'll fix the issues it found.

**Comment by JakeWharton** [comment]
I'm going to steal this change from you, rebase it on master, and see if I can clean up some of the generated code (I don't like the double method dispatch). I'll squash and amend to your changes so you retain the commit author.

**Comment by emilsjolander** [comment]
cool, sounds good!

**Pull request closed without merging**